retrieval/discovery/kubernetes/discovery.go

```diff
@@ -215,11 +229,78 @@ func (kd *Discovery) Run(ch chan<- *config.TargetGroup, done <-chan struct{}) {
 	}
 }
 
+func (kd *Discovery) queryMasterPath(path string) (*http.Response, error) {
+	req, err := http.NewRequest("GET", path, nil)
+	if err != nil {
+		return nil, err
+	}
+	return kd.queryMasterReq(req)
+}
+
+func (kd *Discovery) queryMasterReq(req *http.Request) (*http.Response, error) {
+	// Lock in case we need to rotate masters to request.
+	kd.mastersMu.Lock()
+	defer kd.mastersMu.Unlock()
+	for i := 0; i < len(kd.Conf.Masters); i++ {
+		cloneReq := *req
+		cloneReq.URL.Host = kd.Conf.Masters[0].Host
+		cloneReq.URL.Scheme = kd.Conf.Masters[0].Scheme
+		res, err := kd.client.Do(&cloneReq)
+		if err == nil {
+			return res, nil
+		}
+		kd.rotateMasters()
+	}
+	return nil, fmt.Errorf("Unable to query any masters")
+}
+
+func (kd *Discovery) rotateMasters() {
+	if len(kd.Conf.Masters) > 1 {
+		kd.Conf.Masters = append(kd.Conf.Masters[1:], kd.Conf.Masters[0])
+	}
+}
+
+func (kd *Discovery) updateMastersTargetGroup() *config.TargetGroup {
+	tg := &config.TargetGroup{
+		Source: mastersTargetGroupName,
+		Labels: model.LabelSet{
+			roleLabel: model.LabelValue("master"),
+		},
+	}
+
+	for _, master := range kd.Conf.Masters {
+		masterAddress := master.Host
+		_, _, err := net.SplitHostPort(masterAddress)
+		// If error then no port is specified - use default for scheme.
+		if err != nil {
+			switch master.Scheme {
+			case "http":
+				masterAddress += ":80"
```

Comment:
Please use http://golang.org/pkg/net/#JoinHostPort, as it will take care of bracketing hosts that already contain colons (IPv6, if Kubernetes ever supports that :) ).